### PR TITLE
backup service send in batch

### PR DESCRIPTION
**Pull request by msmouse**
## Description

Looks like hyper::body::Sender::send_data() waits for the ack in place without buffering, that could be the reason why our backup speed is so dependent on network latency.  Upgrading hyper is non-trivial (warp depends on the old version), so I did some manual buffering. 

Doubles the throughput.

image

## Type of Change
- [x] New feature

## Which Components or Systems Does This Change Impact?
- [ ] Other (specify) -- backup service

## How Has This Been Tested?

Existing unit tests (commenting out flush_buffer(), for example, does fail the tests)

patched testnet pfn0

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: storage/backup/backup-service/src/handlers/utils.rs [hunk 1]
```diff
@@ -44,30 +44,115 @@ pub(super) fn reply_with_bcs_bytes<R: Serialize>(
...
+
+        if sender_task.is_finished() {
+            return Err(AptosDbError::Other(
+                "Sender task finished unexpectedly.".to_string(),
+            ));
+        }
+
+        self.buffer.put_slice(bytes);
+
+        const TARGET_BATCH_SIZE: usize = if cfg!(test) { 10 } else { 1024 * 1024 };
```
Comment (areshand) [inline storage/backup/backup-service/src/handlers/utils.rs:115]: move the const to top of the file?

File: storage/backup/backup-service/src/handlers/utils.rs [hunk 2]
```diff
@@ -93,12 +178,17 @@ where
...
-        });
+    match send_size_prefixed_bcs_bytes_impl(iter_res, &mut sender).await {
+        Ok(()) => {
+            if let Err(e) = sender.finish().await {
+                warn!("Failed to finish http body: {:?}", e);
+            }
+        },
+        Err(e) => {
+            warn!("Failed writing http body: {:?}", e);
+            sender.abort().await;
```
Comment (areshand) [inline storage/backup/backup-service/src/handlers/utils.rs:189]: if error happens, we can still use `finish`? any reason for introducing one more abort function?

**Review by areshand — APPROVED** [review]

File: storage/backup/backup-service/src/handlers/utils.rs [hunk 3]
```diff
@@ -44,30 +44,101 @@ pub(super) fn reply_with_bcs_bytes<R: Serialize>(
...
+
+        if sender_task.is_finished() {
+            return Err(AptosDbError::Other(
+                "Sender task finished unexpectedly.".to_string(),
+            ));
+        }
+
+        self.buffer.put_slice(bytes);
+
+        const TARGET_BATCH_SIZE: usize = if cfg!(test) { 10 } else { 1024 * 1024 * 10 };
```
Comment (msmouse) [inline storage/backup/backup-service/src/handlers/utils.rs]: Sorry I deleted your comment accidentally @areshand  :trollface:  

But I like constants to be defined where it's used 🧌

**Review by grao1991 — COMMENTED** [review]
Why only double the throughput, what's the bottleneck now?

**Comment by msmouse** [comment]
@grao1991 some obvious things

1. On the server side the tree iterator without concurrency is like 20k leaves per sec.
2. On the client side, reading from http stream and writing to cloud is not pipelined
3. On the client side, `gzip -c` was like 20MB / s on my machine, `pigz` is 10x that throughput

I'm on 2 as the next step.

**Review by grao1991 — APPROVED** [review]

**Review by areshand — APPROVED** [review]

**Pull request merged**